Add timer tests to make sure event-time timer firing at the right time.

shunping opened the pull request:
**Please** add a meaningful description for your change here

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://github.com/apache/beam/blob/master/CONTRIBUTING.md#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[![Build python source distribution and wheels](https://github.com/apache/beam/actions/workflows/build_wheels.yml/badge.svg?event=schedule&&?branch=master)](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[![Python tests](https://github.com/apache/beam/actions/workflows/python_tests.yml/badge.svg?event=schedule&&?branch=master)](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[![Java tests](https://github.com/apache/beam/actions/workflows/java_tests.yml/badge.svg?event=schedule&&?branch=master)](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[![Go tests](https://github.com/apache/beam/actions/workflows/go_tests.yml/badge.svg?event=schedule&&?branch=master)](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

shunping commented:
It appears that two of the new tests (test_pardo_timer_no_firing, and test_pardo_timer_no_reset) are not working with FnRunner and  PortableRunner. @lostluck

shunping commented:
I think they have the same problem of prism before my last fix: i.e. not supporting timer clearing across bundles.

lostluck commented:
The go to in this instance is check what happens on Dataflow. I'm pretty sure we've implemented things correctly and fixed legitimate bugs in the Prism implementation, but it's also an overall uncommon case, since  windows + OnWindowExpiry is how that's handled usually in Java pipelines IIRC.

I know the Python runner implementations are incomplete in many ways. Not sure where Flink is in this case, and Spark and Samza were further behind.

shunping commented:
> The go to in this instance is check what happens on Dataflow. I'm pretty sure we've implemented things correctly and fixed legitimate bugs in the Prism implementation, but it's also an overall uncommon case, since windows + OnWindowExpiry is how that's handled usually in Java pipelines IIRC.
> 
> I know the Python runner implementations are incomplete in many ways. Not sure where Flink is in this case, and Spark and Samza were further behind.

Yes. I verified that the behavior of prism and dataflow is the same on these tests. The problem is related to FnApiRunner (which is the go-to runner if we specify DirectRunner in python). 

Filed an issue: #35168

shunping commented:
By default, PortableRunner Tests are run against FnApiRunner. That's why the tests on both FnApiRunnerTest and PortableRunnerTest are failing. https://github.com/apache/beam/blob/a96938664a7a7a9ea1b49031b295bce307e6966e/sdks/python/apache_beam/runners/portability/local_job_service.py#L316

shunping commented:
We should be able to re-enable the tests in FnApiRunnerTest and PortableRunnerTest once Prism is set as the default (#34612). (cc'ed @damccorm)

shunping commented:
I have to take out one of the test, because its behavior depends on the runner's bundle size (#35176).
